### DISCUSSION TITLE
2008/03/21/%E8%B0%83%E7%94%A8%E5%8A%A8%E6%80%81%E9%93%BE%E6%8E%A5%E5%BA%93

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#212] [Announcements]
# 2008/03/21/%E8%B0%83%E7%94%A8%E5%8A%A8%E6%80%81%E9%93%BE%E6%8E%A5%E5%BA%93

新版本请查看： https://lv.qizhen.xyz

https://blog.qizhen.xyz/2008/03/21/%E8%B0%83%E7%94%A8%E5%8A%A8%E6%80%81%E9%93%BE%E6%8E%A5%E5%BA%93

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2010年04月22日 12:54 编辑
我从没遇到过这个问题，所以不了解。不过有可能是像它提示的那样，LabVIEW在安装过程中除了些问题，造成一些文件找不到。

回复
影儿
2010年04月22日 11:37 编辑
我在打开一个调用DLL函数的例程时，总会出现下面这样对话框，（因为不能粘贴图片，所以只能描述下），对话框的标题是“LABVIEW.exe-损坏的图像”，报错内容为“应用程序或DLL为无效的Windows映射，请再检查一遍你的安装盘。”请问老师，是什么原因造成的？我把该对话框关闭，该例程还是可以正常使用。我调用DLL的方法是通过配置CLN节点的方法来调用DLL的。 在我电脑上，如果打开一个没调用DLL的例程时，是没有该对话框出现的。

回复
勇卓
2008年11月3日 16:43 编辑
应该是多线程引起的问题，vc的比较容易搞定，labview不清楚。

回复
Qizhen
2008年03月28日 16:55 编辑
浣熊爱上猫，这种 VC linking error，我也不是特别熟悉。最好找个你周围比较精通VC的人跟你一起调试一下。从你的错误信息看，可能是选错了VC的运行库。试着选别的运行库试一下。还可以找论坛去发贴问问。如果你实在解决不了这个问题，我有个同事是解决 linking error 的专家：http://arrowpig1979.spaces.live.com/

回复
๑۩۞۩万年小妖๑۩۞۩
2008年03月28日 12:57 编辑
完蛋了。我统统看不懂…..》。《

回复
浣熊爱上猫
2008年03月25日 14:38 编辑
大师又来打扰了，上次说的返回指针的问题还没有解决，我正在想办法用内嵌c的方式处理数据。但是又遇到问题，我现在用cin创建一个c文件，c在生成lsb时出现错误
/* CIN source file */#include<stdio.h>#include "extcode.h"
MgErr CINRun(int32 *pDaqData);
MgErr CINRun(int32 *pDaqData) {  FILE *fpCsv;     fpCsv = fopen ("DataSingle.csv","w") ;//打开文件DataSingle ，(只写)     fclose(fpCsv);
        return noErr; }
错误是：
Linking…nafxcwd.lib(dllmodul.obj) : error LNK2005: _DllMain@12 already defined in MSVCRT.lib(dllmain.obj)nafxcwd.lib(dllmodul.obj) : warning LNK4006: _DllMain@12 already defined in MSVCRT.lib(dllmain.obj); second definition ignoredlvsbmain.def : error LNK2001: unresolved external symbol GetLVSBHeaderDebug/jj.lib : fatal error LNK1120: 1 unresolved externalsLINK : fatal error LNK1141: failure during build of exports file执行 link.exe 时出错.
jj.dll – 1 error(s), 0 warning(s)
请问是怎么回事啊？

回复
杰
2008年03月22日 17:12 编辑
Shared Liberary 这个工具啊，方便又省心，量大还优惠。我一直用它。这玩意，我看行！哈哈。
不过我确实遇到过出错的时候，我记得是一个DELPHI生成的原始DLL，后来在LabVIEW里面搞不定，又用VC再包装下，重新编译了一个新的DLL才行。呵呵，非常烦。

回复
Lu
2008年03月22日 12:06 编辑
期待一下后续 ：）

回复

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2009年10月11日 11:02 编辑
类似管道的方法，我还真没试过。我估计LabVIEW中可能没有提供类似的功能。

回复
嵘
2009年10月10日 18:31 编辑
把wait until complication 设为 False后，函数就进入CMD不返回了，起不了LV与控制台程序输入输出进行交互的作用。难道在WINDOWS下就实现不了类似管道的功能吗？

回复
Qizhen
2009年10月10日 09:00 编辑
调用一下就关闭是正常的，你调用的程序运行结束，自然就关闭了。如果希望把控制带打开，可以输入命令“cmd”，并且把wait until complication 设为 False。

回复
嵘
2009年10月9日 15:59 编辑
我试过System Exec.vi，但调用一下进程就关闭了。现在想与控制台程序的输入输出进行交互，应该怎么做？

回复
Qizhen
2009年10月9日 11:15 编辑
调用console标准输入输出控制台是非常简单和直观的。使用LabVIEW中的 “Connectivity->Libraries and Executables->System Exec.vi”就可以了。在控制台上输入什么命令，就直接传给这个函数。它返回的文本，就是在控制台上显示的内容。

回复
嵘
2009年10月6日 12:47 编辑
请大师具体讲讲在LABVIEW中如何调用console标准输入输出控制台？

回复
Qizhen
2008年08月27日 19:12 编辑
我的水平和时间都有限，只能侧重写一些自己熟悉的方面。

回复
tscwl
2008年08月27日 13:55 编辑
写的一般，只有基础的东西，应该整点工程实际的

回复

**Comment by ruanqizhen** [comment]
之前博客下的留言：

yi
2008年04月7日 22:28 编辑
太专业了吧，惭愧。。。

回复
WL
2008年04月7日 10:09 编辑
不懂啊

回复
杰
2008年04月6日 16:51 编辑
很有用，不过等待补充和解释，呵呵
譬如说最后一个，string input and output, 是为了告诉LV这个字符串的大小，以便开内存是吧？

回复

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2010年05月17日 14:24 编辑
既然有报错，你就可以按照报错信息进行修改了。

回复
lei
2010年05月17日 09:27 编辑
阮老师，你好！我用LABVIEW作了一个DLL,用labview程序可以调用，但是我用VC++调用，我把DLL文件复制在WINDOWS->SYSTEM下了，在编译时就会报错！请指教！谢谢！

回复
Qizhen
2010年03月9日 13:10 编辑
Delphi的函数参数压栈方式与C不同。你需要查找一些相关资料，也许改变一下参数顺序就好了，但我也不确定，没用过Delphi.

回复
俊俊
2010年03月8日 20:03 编辑
你好，我用Labview调用Delphi编写的一个dll时，有个函数的参数是指向一个结构的指针，我按照你上文中所说的使用一个Cluster来对应，结果运行出现一个严重内部错误：ThEvent.cpp中209行出现错误，请问你遇到过这样的问题吗？能不能给我提供一些解决问题的思路？多谢。

回复
Qizhen
2009年09月10日 09:28 编辑
你的这个想法恐怕实现不了。你自己开发的DLL若需要改变LabVIEW控件中的数据，实现起来相当麻烦，需要调用LabVIEW中一些特殊的函数，而我也不是太确定这些函数是否是公开给外部用户的。比较容易的实现方法是，在LabVIEW中编写一个多线程程序。就是说用动态调用的方式开一个新的线程，这个线程专用于更新控件上的数据。它时刻监视DLL返回的数据，一旦数据有所变化，就去更新控件上的数据。

回复
星
2009年09月8日 17:33 编辑
你好~我有个labview的难道解决了很久没成功.想请教你一下.比如我用labview来做界面,用dll来处理信息.在labview上有些控件我想让dll自主的去更新,而不是由labview的数据流程去更新.比如有一个文本显示控件.我有什么办法可以在初始化时把这个控件告诉dll.然后在以后的某个时刻dll自己去更新了文本显示控件的内容.些时不由labview的执行流程决定…谢谢..有什么好的解决方法麻烦告诉我一下..也可以随便发个邮件给我..email:xtliaohanxin@163.com

回复
guo
2008年04月23日 14:00 编辑
我现在想调用dll库中的函数，函数设置是long LEDAPI AddText(char *str, LPRECT rect, long method, long speed, long transparent, char *fontname, long fontsize, long fontcolor); 现在想知道str  rect在labview中的数据类型是？查了好久，还是不明白怎么设置，谢谢大侠啦

回复
arrowpig
2008年04月12日 21:44 编辑
这个工具小牛牛还在NI 10周年庆的时候说道了。确实啊，蛮多人的心血在里面的。最最开始的时候丁立超整天在唠叨，说CLN太难用了，很多客户打电话给他，他都累死了，到现在做的比较完善了。

回复
杰
2008年04月11日 17:41 编辑
相当有用，已收藏，谢了

回复

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2009年05月18日 21:05 编辑
LabVIEW不能回收，若C语言开辟了内存，则必须提供销毁内存的函数，LabVIEW调用该函数销毁其内存。

回复
Kun
2009年05月18日 17:44 编辑
请问在外部C语言中分配的一块内存，labview程序是怎么收回的？

回复
YALU
2008年05月5日 01:28 编辑
最近怎样? 我很久没有上网了! 终于复活了!

回复
竹贤
2008年05月4日 21:13 编辑
请你简要介绍一下这几个VI吧。

回复
mnihj
2008年05月4日 11:31 编辑
这个模块太有用了，以后不用再因为返回指针而费尽心思了。

回复
